File: BE/SSAIREN/src/main/java/com/ssairen/common/dto/ApiResponse.java
package com.ssairen.common.dto;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonInclude;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;

import java.time.ZonedDateTime;

/**
 * 통일된 API 응답 형식
 * @param <T> 응답 데이터 타입
 */
@Getter
@Builder
@AllArgsConstructor
@JsonInclude(JsonInclude.Include.NON_NULL)
public class ApiResponse<T> {

    private boolean success;
    private T data;
    private String message;
    private ErrorResponse error;
    private Integer status;

    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ssXXX")
    private ZonedDateTime timestamp;

    /**
     * 성공 응답 생성 (메시지 포함)
     */
    public static <T> ApiResponse<T> success(T data, String message) {
        return ApiResponse.<T>builder()
                .success(true)
                .data(data)
                .message(message)
                .timestamp(ZonedDateTime.now())
                .build();
    }

    /**
     * 성공 응답 생성 (메시지 없음)
     */
    public static <T> ApiResponse<T> success(T data) {
        return success(data, null);
    }

    /**
     * 성공 응답 생성 (데이터 없음)
     */
    public static ApiResponse<Void> success(String message) {
        return ApiResponse.<Void>builder()
                .success(true)
                .message(message)
                .timestamp(ZonedDateTime.now())
                .build();
    }

    /**
     * 에러 응답 생성
     */
    public static <T> ApiResponse<T> error(ErrorResponse error, int status) {
        return ApiResponse.<T>builder()
                .success(false)
                .error(error)
                .status(status)
                .timestamp(ZonedDateTime.now())
                .build();
    }
}

File: FE/ssairen_app/app/src/main/java/com/example/ssairen_app/data/dto/ApiResponse.kt
package com.example.ssairen_app.data.dto

import com.google.gson.annotations.SerializedName

/**
 * 백엔드 API 공통 응답 형식
 */
data class ApiResponse<T>(
    @SerializedName("success")
    val success: Boolean,

    @SerializedName("data")
    val data: T? = null,

    @SerializedName("message")
    val message: String? = null,

    @SerializedName("error")
    val error: ErrorResponse? = null,

    @SerializedName("status")
    val status: Int? = null,

    @SerializedName("timestamp")
    val timestamp: String? = null
)

/**
 * 에러 응답
 */
data class ErrorResponse(
    @SerializedName("code")
    val code: String,

    @SerializedName("message")
    val message: String,

    @SerializedName("details")
    val details: List<ErrorDetail>? = null
)

/**
 * 에러 상세
 */
data class ErrorDetail(
    @SerializedName("field")
    val field: String,

    @SerializedName("message")
    val message: String,

    @SerializedName("rejectedValue")
    val rejectedValue: Any?
)
